### PR TITLE
Improve grid data providers

### PR DESCRIPTION
**Pull request by sarjon**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | This PR aims to: 1. make pagination & sorting optional in `SearchCriteria`. This allows to use same data providers to export data when needed. 2. Introduces new hook to allow grid query modifications for modules.
| Type?         | improvement
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | n/a
| How to test?  | Pages with grid should work as before.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/9402)

### CODE REVIEW

**Comment by sarjon** [comment]
@mickaelandrieu wdyt?

File: src/PrestaShopBundle/Entity/Repository/LogRepository.php
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Entity/Repository/LogRepository.php]: As this function you introduce will be only used in repositories, you'd better use an abstract class.

File: src/Core/Grid/DataProvider/DoctrineGridDataProvider.php
Comment (mickaelandrieu) [inline src/Core/Grid/DataProvider/DoctrineGridDataProvider.php]: note to myself: we need an interface for this.

**Review by mickaelandrieu — COMMENTED** [review]
maybe use an abstract class instead of a trait.

File: src/PrestaShopBundle/Entity/Repository/LogRepository.php
Comment (mickaelandrieu) [reply src/PrestaShopBundle/Entity/Repository/LogRepository.php]: ping @eternoendless, wdyt? How do you decide when to use an abstract class or a trait?

Comment (sarjon) [reply src/PrestaShopBundle/Entity/Repository/LogRepository.php]: in this case, trait does not provide any public API and it does not have any dependencies. So you can use it in both repositories or other services responsible for building queries. 

p.s. @mickaelandrieu i know you dont like traits. :smiley:

Comment (mickaelandrieu) [reply src/PrestaShopBundle/Entity/Repository/LogRepository.php]: It's not about like or not like, but I'm not sure we should use this trait "outside" of repositories.

For instance, I'm easy with something like TranslatorTrait as I always change the signature of `trans` function to make domain the second argument. As translations are needed everywhere and on a lot of unrelated classes, I guess this is a valid use case for a trait.

I'd like the opinion of @eternoendless or @PierreRambaud or anyone form @PrestaShop/prestashop-core-dev here :+1:

Comment (sarjon) [reply src/PrestaShopBundle/Entity/Repository/LogRepository.php]: not exactly to the topic, but i did use query building outside for repository here https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Grid/Query/EmailLogsQueryBuilder.php do you think it should be updated to repository?

Comment (eternoendless) [reply src/PrestaShopBundle/Entity/Repository/LogRepository.php]: I'm not a fan of traits. Some define them as "language-assisted copy-paste" – which in all fairness is pretty much true.

While they may seem useful at first sight, traits are really treacherous. They can be too easily misused to violate the Single Responsibility principle, because they provide the _illusion_ of clean design and composition. But make no mistake, when you add a trait, you are actually _appending_ behavior to your class. It just happens to be in another file, hidden from view.

So whenever you are tempted to add a trait, ask yourself: "are the methods in this trait part of the class' responsibility?" Most of the time they will not, so you can get it done using real composition (ie. injecting another object) instead.

Comment (sarjon) [reply src/PrestaShopBundle/Entity/Repository/LogRepository.php]: @eternoendless in this case method provided by trait is this class responsibility. This method does not provide public API for class.

Reason why i added this trait is that i noticed that all query builder have exactly the same few lines of code (adding sorting & pagination to doctrine queries). 

what do you think?

Comment (eternoendless) [reply src/PrestaShopBundle/Entity/Repository/LogRepository.php]: This is reusable logic that I think it would make more sense as a separate class. To me, it looks like it enriches a QueryBuilder using SearchCriteriaInterface. As a bonus, you will be able to add unit test it (try it!).

Think of it as a class whose role is to enrich a query (any query) using SearchCriteria. 

Usage would be like this:

```php
$this->getSearchCriteriaApplicator()
    ->applySorting($searchCriteria, $qb)
    ->applyPagination($searchCriteria, $qb);
```

Following this train of though, I think the `buildGridQuery` method could be even further enhanced.

Check out this block:

```php
$filters = $searchCriteria->getFilters();
foreach ($filters as $filterName => $filterValue) {
    if (empty($filterValue)) {
        continue;
    }

    if ('employee' == $filterName) {
        $qb->andWhere('e.lastname LIKE :employee OR e.firstname LIKE :employee');
        $qb->setParameter('employee', '%'.$filterValue.'%');
        continue;
    }

    if ('date_add' == $filterName) {
        if (!empty($filterValue['from']) &&
            !empty($filterValue['to'])
        ) {
            $qb->andWhere('l.date_add BETWEEN :date_from AND :date_to');
            $qb->setParameters(array(
                'date_from' => $filterValue['from'],
                'date_to' => $filterValue['to'],
            ));
        }
        continue;
    }

    $qb->andWhere("$filterName LIKE :$filterName");
    $qb->setParameter($filterName, '%'.$filterValue.'%');
}
```

There are two problems with it.

First, [Switch Is Bad](https://sourcemaking.com/refactoring/smells/switch-statements). Same goes for a sequence of `if`. Think about this: How would you add another filter? By editing the file? How would _somebody else_ do it?.

Second, this is a typical example of procedural micro-management, as opposed to the ["Tell, don't ask" principle](https://pragprog.com/articles/tell-dont-ask):

> Procedural code gets information then makes decisions. Object-oriented code tells objects to do things. 
> — Alec Sharp

The first thing to do is identifying our actors:

- We have a `SearchCriteriaInterface` which defines a couple known criteria (`orderBy`, `orderWay`, `offset` and `limit`) as well a number of unknown criteria (`filters`).
- Then there is a `QueryBuilder` where we need to apply that criteria.

Currently, the knowledge on how to match that `SearchCriteria` to the `QueryBuilder` is hardcoded in the `Repository`.

This can be extracted using a `SearchCriteriaApplicator`, and replace the whole block with this:

```php
$this->getSearchCriteriaApplicator()->apply($searchCriteria, $qb);
```

Now that looks simpler! We simply extracted the previous code and put it inside a `SearchCriteriaApplicator::apply()` method. This solves the "tell, don't ask" issue and makes code easier to understand.

But we still have to ugly, hard-to-extend sequence of `if` to refactor. How can we improve on that?

First, let's extract each `if` case into its own method:

```php
// class SearchCriteriaApplicator

public function apply(SearchCriteriaInterface $searchCriteria, QueryBuilder $qb)
{
    $filters = $searchCriteria->getFilters();
    foreach ($filters as $filterName => $filterValue) {
        if (empty($filterValue)) {
            continue;
        }
    
        if ($this->applyEmployee($filterName, $filterValue $qb)) {
            continue;
        }
        
        if ($this->applyDateAdd($filterName, $filterValue, $qb)) {
            continue;
        }
    
        $this->applyOther($filterName, $filterValue, $qb);
    }
}

// [...]

private function applyEmployee($filterName, $filterValue, QueryBuilder $qb)
{
    if ('employee' == $filterName) {
        $qb->andWhere('e.lastname LIKE :employee OR e.firstname LIKE :employee');
        $qb->setParameter('employee', '%'.$filterValue.'%');
        return true;
    }
    return false;
}

private function applyDateAdd($filterName, $filterValue, QueryBuilder $qb)
{
    if ('date_add' == $filterName) {
        if (!empty($filterValue['from']) &&
            !empty($filterValue['to'])
        ) {
            $qb->andWhere('l.date_add BETWEEN :date_from AND :date_to');
            $qb->setParameters(array(
                'date_from' => $filterValue['from'],
                'date_to' => $filterValue['to'],
            ));
        }
        return true;
    }
    return false;
}

// [...]
```

Now, if we refactor it a little further and extract previously extracted methods into classes of their own, we can stop looping on the the `$filters` and loop on `FilterApplicators` instead:

```php
// class SearchCriteriaApplicator

/** @param FilterApplicatorInterface[] $criteriaApplicator */
public function __construct(array $criteriaApplicators);

public function apply(SearchCriteriaInterface $searchCriteria, QueryBuilder $qb)
{
    $filters = $searchCriteria->getFilters();
    if (!empty($filters) {    
        foreach ($this->getCriteriaApplicators() as $applicator) {
            $applicator->apply($filters, $qb);
        }
    }
}

// class EmployeeApplicator

private function match($filters)
{
    return isset($filters['employee']) && !empty($filters['employee']);
}

public function apply(array $filters, QueryBuilder $qb)
{
    if ($this->match($filters) {
        $qb->andWhere('e.lastname LIKE :employee OR e.firstname LIKE :employee');
        $qb->setParameter('employee', '%'.$filter['employee'].'%');
    }
}

// class DateAddApplicator

private function match($filters)
{
    if (!isset($filters['date_add']) {
        return false;
    }
    $filterValue = $filters['date_add'];
    
    return !empty($filterValue['from']) && !empty($filterValue['to']);
}

public function apply(array $filters, QueryBuilder $qb)
{
    if ($this->match($filters) {
        $filterValue = $filters['date_add'];
        
        $qb->andWhere('l.date_add BETWEEN :date_from AND :date_to');
        $qb->setParameters(array(
            'date_from' => $filterValue['from'],
            'date_to' => $filterValue['to'],
        ));
    }
    return true;
}
```

This way, we are inversing responsibility, and making applicators _subscribe_ to the presence of a filter on a filter list instead of calling them explicitely. To add a new filter, you only need to add a new `FilterApplicator` on the list contained by `SearchCriteriaApplicator`, which can of course be done in runtime (no more file editing!).

Comment (mickaelandrieu) [reply src/PrestaShopBundle/Entity/Repository/LogRepository.php]: As I totally agree with what you're saying @eternoendless, I have the feeling the refactoring you are suggesting should be done in another pull request to make the scope of this one limited.

How about creating an issue into prestafony project and set a milestone to ensure this will be done during 1.7.5 "time of development" ? :+1: 

> you can copy/paste your comment which is really clear about what you want us to do!

Comment (sarjon) [reply src/PrestaShopBundle/Entity/Repository/LogRepository.php]: @eternoendless as usual, great explanation and i really like your idea!

How would you suggest dealing with aliases in SQL? For example:

```php
public function apply(array $filters, QueryBuilder $qb)
{
    if ($this->match($filters) {
        $filterValue = $filters['date_add'];
        
        $qb->andWhere('l.date_add BETWEEN :date_from AND :date_to');
        $qb->setParameters(array(
            'date_from' => $filterValue['from'],
            'date_to' => $filterValue['to'],
        ));
    }
    return true;
}
```
And in this line ` $qb->andWhere('l.date_add BETWEEN :date_from AND :date_to');` we have `l.date_add`,  but in other SQL it may be `a.date_add` or `b.date_add`.

Would we need different applicators for this? Or maybe we should try to avoid alias and just keep ` $qb->andWhere('date_add BETWEEN :date_from AND :date_to');` so applicator would be reusable?

Comment (eternoendless) [reply src/PrestaShopBundle/Entity/Repository/LogRepository.php]: @mickaelandrieu 
> I have the feeling the refactoring you are suggesting should be done in another pull request

Absolutely. The second-level refactoring (which extracts code from `buidGridQuery()`) can be done in a separate PR.

However, the first one (`$this->getSearchCriteriaApplicator()->applySorting($searchCriteria, $qb)->applyPagination($searchCriteria, $qb)`) I'd like it to be done in this PR so that we can get rid of the trait.

@sarjon 
> How would you suggest dealing with aliases in SQL?

That's a very good question. I can think of two ways of solving it: either architecture-based or conventions-based.

In the architecture-based solution, we would have to think of a way of providing Applicators with meta information about how tables are aliased. This information would be provided on construction and used to match a known table to the alias it happens to have in the query builder. While this would may solve our immediate problem, it would also complexity to the whole system.

After a closer look, it looks like it would also be possible to figure out the alias by inspecting the query builder:

```php
$qb->getPart('from')
// contains a collection of array('table' => $tableName, 'alias' => $alias')
```

Of course, also "smells" because of added complexity. Also, aliases may be defined by `JOIN` which would add further complexity.

In the conventions-based solution, we accept that Applicators are closely coupled to the queries they change, and as such they may share implicit (aka. "hard-coded") knowledge. This will probably mean having two Applicators that do pretty much the same thing but with slightly different code because of aliases. If you worry about aliases, then use full table names everywhere.

No solution is perfect, both have tradeoffs.

**Comment by mickaelandrieu** [comment]
Ping @eternoendless, I need this to be merged in order to try to break my own [pull request](https://github.com/PrestaShop/PrestaShop/pull/9324) (because for now it's impossible as people don't have access to the query builder, so they can't add both positional and named parameters in one query)

**Comment by sarjon** [comment]
@eternoendless @mickaelandrieu please review. :pray:

File: src/Core/Grid/DataProvider/DoctrineGridDataProvider.php
Comment (eternoendless) [inline src/Core/Grid/DataProvider/DoctrineGridDataProvider.php]: I think hooks are typically prepended with "action", aren't they?

https://devdocs.prestashop.com/1.7/modules/concepts/hooks/#naming-scheme

File: src/Core/Grid/Query/DoctrineSearchCriteriaApplicator.php
Comment (eternoendless) [inline src/Core/Grid/Query/DoctrineSearchCriteriaApplicator.php]: Not sure if it's really necessary to use the word "Doctrine" everywhere

Comment (mickaelandrieu) [reply src/Core/Grid/Query/DoctrineSearchCriteriaApplicator.php]: We try to keep the opportunity to remove Doctrine at some point: this probably won't happens, but this make the grid component not hardly bounded to the ORM :+1:

File: src/Core/Grid/DataProvider/DoctrineGridDataProvider.php
Comment (sarjon) [reply src/Core/Grid/DataProvider/DoctrineGridDataProvider.php]: well, i guess you are right, although there are some hooks that does not respect this rule. :/ would `actionModifyGridQueryBuilder` be better?

Regarding you comment about applicators, i think that this hook will not be needed anymore, as we could allow attaching custom applicators instead?

File: src/Core/Grid/Query/DoctrineSearchCriteriaApplicator.php
Comment (sarjon) [reply src/Core/Grid/Query/DoctrineSearchCriteriaApplicator.php]: @eternoendless i wanted to make clear separation as we could have different applicators for search criteria. For example, backups list data is loaded from filesystem so we dont touch doctrine there at all.

File: src/Core/Grid/DataProvider/DoctrineGridDataProvider.php
Comment (mickaelandrieu) [reply src/Core/Grid/DataProvider/DoctrineGridDataProvider.php]: we need a hook to be able to attach custom applicators instead then :-)

And this should remains easy to use it

**Comment by sarjon** [comment]
is there anything still missing for this PR? ^^

**Comment by mickaelandrieu** [comment]
> is there anything still missing for this PR? ^^

It's up to @eternoendless to decide as we will apply his comments everywhere :)

File: src/Core/Grid/DataProvider/DoctrineGridDataProvider.php
Comment (mickaelandrieu) [inline src/Core/Grid/DataProvider/DoctrineGridDataProvider.php]: here, how can you manage to change the grid query builder for a specific Grid? :)

Comment (sarjon) [reply src/Core/Grid/DataProvider/DoctrineGridDataProvider.php]: Via injected grid_id, of course it duplicates grid Id in definition factory `getId()`  method and data provider service definition. :/

Comment (mickaelandrieu) [reply src/Core/Grid/DataProvider/DoctrineGridDataProvider.php]: you know how you can update this hook name ^_^ => https://github.com/PrestaShop/docs/pull/116

**Comment by sarjon** [comment]
rebased.

File: src/Core/Grid/DataProvider/DoctrineGridDataProvider.php
Comment (sarjon) [reply src/Core/Grid/DataProvider/DoctrineGridDataProvider.php]: i guess i should remove it as its done in our PR?

Comment (mickaelandrieu) [reply src/Core/Grid/DataProvider/DoctrineGridDataProvider.php]: yes :)

Comment (sarjon) [reply src/Core/Grid/DataProvider/DoctrineGridDataProvider.php]: ill update it tomorrow morning, is there anything else to do regarding grid? tests, docs?

Comment (mickaelandrieu) [reply src/Core/Grid/DataProvider/DoctrineGridDataProvider.php]: hmm ... i've just merged https://github.com/PrestaShop/docs/pull/116, do you think you can start to document every column type for instance?

I'll update architecture and docs for architecture soon (after merge of https://github.com/PrestaShop/PrestaShop/pull/9437, then  https://github.com/PrestaShop/PrestaShop/pull/9470, then this one, then ... xD)

Comment (sarjon) [reply src/Core/Grid/DataProvider/DoctrineGridDataProvider.php]: and then https://github.com/PrestaShop/PrestaShop/pull/9402 :D 

Ok, so ill work on Columns type docs, i also have some docs regarding DataFactory, so ill see if i can update it. :)

**Comment by mickaelandrieu** [comment]
Hello @sarjon, this one is the next and last one before freeze imo :+1:

File: src/Core/Grid/DataProvider/DoctrineGridDataProvider.php
Comment (mickaelandrieu) [reply src/Core/Grid/DataProvider/DoctrineGridDataProvider.php]: edit: done and merged :+1:  you can now rely on HookDispatcherInterface

**Comment by sarjon** [comment]
rebased :+1:

File: src/Core/Grid/Data/Factory/DoctrineGridDataFactory.php
Comment (sarjon) [inline src/Core/Grid/Data/Factory/DoctrineGridDataFactory.php]: i guess we dont need this param, right? or if we keep it, we can remove grid id from hook name, that would solve issue with definition id having `_`?

Comment (mickaelandrieu) [reply src/Core/Grid/Data/Factory/DoctrineGridDataFactory.php]: ? we need the definition Id in the hook, or your hook will impact ALL grids at the same time

Comment (sarjon) [reply src/Core/Grid/Data/Factory/DoctrineGridDataFactory.php]: not really, you would have to check for grid id in your hook, similar approach is done it some other hooks.

Comment (mickaelandrieu) [reply src/Core/Grid/Data/Factory/DoctrineGridDataFactory.php]: I don't want that, I want a specific hook for each grid. If you need to filter the conditions of using a hook, it's already a defeat regarding the developer experience.

Comment (mickaelandrieu) [reply src/Core/Grid/Data/Factory/DoctrineGridDataFactory.php]: please stick to the documented schema, and tell me if something is not possible :)

Comment (sarjon) [reply src/Core/Grid/Data/Factory/DoctrineGridDataFactory.php]: ok, then lets remove `grid_id` param. :+1:

**Comment by sarjon** [comment]
rebased (again). :+1:

**Review by mickaelandrieu — DISMISSED** [review]

**Comment by marionf** [comment]
Hello @sarjon 

I have an exception in the emails page with this PR

capture d ecran_248

**Comment by sarjon** [comment]
@marionf sorry, fixed now.

**Comment by marionf** [comment]
Thank you @sarjon :+1:

**Comment by PierreRambaud** [comment]
@sarjon I rerun travis, wait for CI and merge :)

**Comment by mickaelandrieu** [comment]
And ... this one is the last feature we accept in the Grid component: time to complete docs and tests, next improvements will wait for 1.7.6.

Good job everyone, we did it: we made a complete and highly configurable Grid component in less than 3 months :tada: !

**Review by PierreRambaud — APPROVED** [review]

**Comment by PierreRambaud** [comment]
Thanks @sarjon and everyone for reviews ;)

**Pull request merged**
